Initial MDRAID support, F2FS documentation

Harvie opened the pull request:
This allows to create level 1 MDRAID with 1 device. Can be empty or prepopulated with data image.
Includes unit test using mdadm to check generated image.

Eg.:

```ini
image mdraid-md.img {
  mdraid {
    level = 1
  }
  partition data {
    image = "mdraid-ext4.img"
  }
}
```

It might sound stupid to create single device raid, but it actualy fits my reallife usecase where i pre-generate such raid when making image and user then can very easily add more devices during runtime when needed later. For example like this:

```sh
mdadm --grow /dev/md23 --raid-disks=2 --force
maddm /dev/md23 --add /dev/sdb1
```

Although it should be simple to pre-generate images for all raid members in genimage. It shouldn't be much harder than generating multiple images with the same RAID UUID and few metadata changes. But i haven't researched that so far.

Also see #191

Harvie commented:
I have no idea why the test fails here. It says file `mdraid.config` is missing, but it's obviously there.
Also tests are passing at my PC when i do `make check-TESTS`:

![image](https://github.com/user-attachments/assets/455cc966-6e2a-4c46-a362-5f6ed10cc68e)

michaelolbrich commented:
https://github.com/pengutronix/genimage/actions/runs/12428468239/job/34700133730?pr=277#step:6:1000

The test run `make distcheck` which runs the test from the generated tarball. You forgot to add the test config to `EXTRA_DIST` so it's missing.

Harvie commented:
Seems to be fixed now.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+		newcsum += __le16_to_cpu(*(unsigned short*) isuper);
+
+	csum = (newcsum & 0xffffffff) + (newcsum >> 32);
+	sb->sb_csum = disk_csum;
```
This is redundant. `sb_csum` is immediately overwritten by the return value of this function. Just set the new value directly here.

Harvie replied:
not really, calling the function does not imply you want to set the checksum in the metadata. if you read carefuly you will see it temporarily sets the checksum to 0 and then sets it back to original value (rather than calculated one). This is done because checksum in metada has to be 0 when calculating checksum of metadata, otherwise that's gonna be endless recursion.

Harvie replied:
In future i might need to calculate crc without overwriting it. I've already usecase for that.
(generating member disk for pre-existing array that was not generated by genimage)

michaelolbrich replied:
I don't understand that. What do you need the old invalid crc for?

Harvie replied:
It's function that calculates CRC. Not function that calculates AND sets CRC. It's more useful, because it can be used for both. Populating metadata with correct CRC and validating CRC of existing metadata loaded from disk. The version you propose is only useful for creating new metadata. my version is good for both.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	struct mdp_superblock_1 *sb = xzalloc(sizeof(struct mdp_superblock_1)+max_devices*2);
+
+	/* constant array information - 128 bytes */
+	sb->magic = 0xa92b4efc;		/* MD_SB_MAGIC: 0xa92b4efc - little endian */
```
create a macro for the magic, even though it's only used once.

And put the comments above each assignment in a separate line

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	sb->feature_map = 0; //MD_FEATURE_BITMAP_OFFSET;	/* bit 0 set if 'bitmap_offset' is meaningful */ //TODO: internal bitmap bit is ignored, unless there is correct bitmap with BITMAP_MAGIC in place
+	sb->pad0 = 0;		/* always set to 0 when writing */
+
+	random_uuid(sb->set_uuid);	/* user-space generated. U8[16]*/ //TODO: should we allow user to set this?
```
Add an option for this please. It should be possible to make image generation reproducible.

Harvie replied:
to be honest. i was trying to avoid complexity of UUID parsing and handling, but now i've noticed there seems to be comprehensive toolkit in util.c that i can use :-)

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+
+	random_uuid(sb->set_uuid);	/* user-space generated. U8[16]*/ //TODO: should we allow user to set this?
+	strncpy(sb->set_name, name, 32); sb->set_name[31]=0;	/* set and interpreted by user-space. CHAR[32] */
+	sb->ctime=time(NULL);	/* lo 40 bits are seconds, top 24 are microseconds or 0*/
```
1. spaces around `=` (the same everywhere else)
2. `& 0xffffffffff` to make the lower 40 bit explicit
3. add an options for this for reproducibility

Harvie replied:
1 & 2 done, 3 pending

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	    offsetof(struct mdp_superblock_1, utime) != 192 ||
+	    sizeof(struct mdp_superblock_1) != 256) {
+		fprintf(stderr, "WARNING - superblock isn't sized correctly\n");
+	}
```
Why is this commented out?

Harvie replied:
> And put it directly after the struct definition. 

that struct is in #include <linux/raid/md_p.h> so i don't think that's really feasible...

Harvie replied:
I'll put it just after definitions.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+
+	sb->dev_number=0;	/* permanent identifier of this  device - not role in raid */
+	sb->cnt_corrected_read=0; /* number of read errors that were corrected by re-writing */
+	random_uuid(sb->device_uuid); /* user-space setable, ignored by kernel U8[16] */ //TODO: should we allow user to set this?
```
Same here: add an option.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+
+	int raid_level = cfg_getint(image->imagesec, "level");
+
+	if(raid_level != 1) {
```
Space after `if` everywhere.

Harvie replied:
done

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+			if(part->image) {
+				image_info(image, "MDRAID using data from [%s]: %s\n", part->name, part->image);
+				img_in = image_get(part->image);
+				if(image->size < (img_in->size + DATA_OFFSET_BYTES)) image->size = (img_in->size + DATA_OFFSET_BYTES);
```
No. Don't overwrite the size silently. Set it if the size is zero. If it's too small, fail with an error.

And put the assignment in a new line.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	int align = 8*512;
+	if(image->size % align) {
+		image->size += align-(image->size % align);
+	}
```
`roundup()` from `image-hd.c` should to just that. Move it to util.h and use it.

Also, only do this, when setting the size above. And fail with an error here, if the size is not a multiple of 4k.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	ret = insert_data(image, sb, imageoutfile(image), sizeof(struct mdp_superblock_1)+max_devices*2, 8*512);
+	if(ret) return ret;
+	if(img_in) {
+		ret = insert_image(image, img_in, image->size-DATA_OFFSET_BYTES, DATA_OFFSET_BYTES, 0);
```
Why not img_in->size for the size argument?

Harvie replied:
because img_in might be smaller than raid itself... i think it's valid use case. eg.: i download small premade-image from internet, but i know i want my partition to be bigger.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+
+	/* constant this-device information - 64 bytes */
+	sb->data_offset=DATA_OFFSET_SECTORS;	/* sector start of data, often 0 */
+	sb->data_size=(image->size - DATA_OFFSET_BYTES)/512;	/* sectors in this device that can be used for data */
```
```
sb->data_size = image->size / 512 - sb->data_offset;
```
That makes it more obvious that the two values are related.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+	int ret;
+	ret = prepare_image(image, image->size);
+	if(ret) return ret;
+	ret = insert_data(image, sb, imageoutfile(image), sizeof(struct mdp_superblock_1)+max_devices*2, 8*512);
```
Remember the size during allocation and reuse it here.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+
+	char *name = cfg_getstr(image->imagesec, "label");
+
+	struct mdp_superblock_1 *sb = xzalloc(sizeof(struct mdp_superblock_1)+max_devices*2);
```
Spaces around '+' here and everywhere else.

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,226 @@
+			image_info(image, "MDRAID partition has to be called 'data' instead of '%s'\n", part->name);
+		} else {
+			if(img_in) {
+				image_error(image, "MDRAID cannot contain more than one data partition!\n");
```
There will never be more than one, right? In that case, take a look at how `image-android-sparse.c` does this:
Add an single option and create the partition internally.

Harvie replied:
i was thinking that in future i might use this to reference to other member drives of the same RAID array. in that case there will be other "partitions" in the config.

Harvie replied:
I think it makes sense tho, therefore i've fixed it as suggested.

Harvie replied:
Actualy there seems to be downside of this. image with file handler is not implicitely created this way, which means we need to manualy create `image something.img { file {} }` in order for something.img to be useable as image. Not sure what would be the proper way to handle this... i suspect `image-android-sparse.c` has the same issue.

Harvie replied:
Nevermind. Trick was to register the partition in parse() before using it in setup(). I was only using setup(), which was the problem...

michaelolbrich commented on test/misc.test:
```diff
@@ -104,6 +104,14 @@ test_expect_success fiptool "fip" "
+exec_test_set_prereq mdadm
+test_expect_success mdadm "mdraid" "
+	run_genimage_root mdraid.config test.mdraid &&
+	mdadm --examine images/test.mdraid &&
```
Save the output here in a file, and just grep the file below.

Hmm, or better use `test_cmp` to compare the output with a file. With fixed uuids and timestamp, the output should be reproducible.

Harvie replied:
I guess we can do that, but i didn't really wanted to be dependend on exact MDADM version... Hence checking individual fields in the output rather than comparing whole thing, so it keeps working in case they add field or something...

Harvie replied:
yeah. i've tried this and it does not work. mdadm output is not consistent even between both ubuntu versions used in pipeline...

Harvie commented:
Happy holidays.
@michaelolbrich I think i've resolved most of your objections, but i think i could really use your guidance about the multi-image thing. While i am not gonna implement generating multi device raid now in the first mdraid PR, i think it's bound to happen in the future and i would like to prepare the codebase for it, so we don't have to change config syntax in the future. What do you think would be the correct way of doing that?

It would kinda make sense for the image-mdraid.c to actualy generate two (or more) images, all being part of that RAID1 mirror. But on the other hand i don't think this would allow genimage to resolve dependencies correctly. Maybe i should manualy create other image structs and add them to list in mdraid_setup() ? would that be enough for image-hd to find them and trigger the build of all of them when needed? what would you suggest?

Harvie commented:
Maybe i should just do something like

```ini
image raid1-a.img {
  mdraid {
    level = 1
  }
  partition data {
    image = "mdraid-ext4.img"
  }
}

image raid1-b.img {
  mdraid {
    master = raid1-a.img
  }
}
```

and then B image can lookup all the configuration from A image config node... but still this does not guarantee that B will be only generated after A. (UUIDs and other details of the A image need to be decided before generating B)

Harvie commented:
OK, i did binary diff of superblocks of two disks belonging to same RAID1 and highlighted the important parts:

![image](https://github.com/user-attachments/assets/efe9e765-8947-4aa1-9853-5ac1aecb9d30)

Some things i am confused about:

- Dev numbers 0 and 2 (why skipped 1? maybe the first disk should be 1, not 0?)
- Why is there 3rd role with 0x0100 ???? Maybe the system is really confused about device ids starting at 0

Harvie commented:
I am now able to create two partitions belonging to single raid like this:

```ini
image mdraid-hd.img {

  hdimage {
    partition-table-type = "gpt"
  }

  partition mdraid-a {
    image = "mdraid-a.img"
    partition-type-uuid = R
  }

  partition mdraid-b {
    image = "mdraid-b.img"
    partition-type-uuid = R
  }

}

image mdraid-a.img {
	mdraid {
		devices = 2
		role = 0
		timestamp = 638022222
		raid-uuid = "de9980f1-0449-4e83-84bd-98e4b1ca3fe3"
	}
	partition data {
		image = "mdraid-ext4.img"
	}
}

image mdraid-b.img {
	mdraid {
		devices = 2
		role = 1
		timestamp = 638022222
		raid-uuid = "de9980f1-0449-4e83-84bd-98e4b1ca3fe3"
	}
	partition data {
		image = "mdraid-ext4.img"
	}
}

image mdraid-ext4.img {
  ext4 {
    label = "TEST_FS"
  }
  size = 5M
}
```

When i do `losetup -fP mdraid-hd.img` the kernel automaticaly recognizes both partitions as members of the array and assembles it automaticaly, while reporting the filesystem in the raid to be clean:

```
# cat /proc/mdstat 
Personalities : [raid1] 
md127 : active raid1 loop0p1[0] loop0p2[1]
      5120 blocks super 1.2 [2/2] [UU]
      
unused devices: <none>

# LANG=C fsck.ext4 /dev/md127 
e2fsck 1.47.1 (20-May-2024)
TEST_FS: clean, 12/1280 files, 1434/5120 blocks
```

Harvie commented:
Great news. With last version the required metadata fields are automaticaly exchanged and synced between components.

```ini
image mdraid-hd.img {

  hdimage {
    partition-table-type = "gpt"
  }

  partition mdraid-a {
    image = "mdraid-a.img"
    partition-type-uuid = R
  }

  partition mdraid-b {
    image = "mdraid-b.img"
    partition-type-uuid = R
  }

}

image mdraid-a.img {
	mdraid {
		level = 1
		devices = 2
		image = "mdraid-ext4.img"
	}
}

image mdraid-b.img {
	mdraid {
		inherit = "mdraid-a.img"
	}
}

image mdraid-ext4.img {
  ext4 {
    label = "TEST_FS"
  }
  size = 5M
}
```

You only need to specify array format for one of the images and other images can refer to it by setting `inherit = "file.img"` config option.

Harvie commented:
I think i am ready here. I've implemented all the features i can wish for. Added test, documented everything. And also added documentation for f2fs, which is something i was kinda owed since my f2fs PR was merged :angel:

michaelolbrich commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+		}
+
+		sb->level = 1;		/* -4 (multipath), -1 (linear), 0,1,4,5 */
+		//sb->layout = 2;		/* only for raid5 and raid10 currently */
```
This is confusing. Is the value a don't care for other raid levels? In that case, maybe use
```
// sb->layout = ignored;
```
Or something else that would not compile.

Harvie replied:
Can do that, BUT. Are we planning to (un)comment random lines? I mean... the code surely is not gonna work if you comment the line before that, but still will compile. Where do we draw the line here?

Harvie replied:
BTW the array is going to be completely valid and fine if you uncomment that line, since every implementation knows to ignore this field for RAID 1. What risk exactly are you trying to avoid?

michaelolbrich replied:
I'm trying to avoid confusion. Where is the '2' coming from?

michaelolbrich reviewed (changes requested):
Mostly cosmetic stuff left.

And please squash all the mdraid commits.

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+#define BITMAP_SECTORS_MAX	256
+#define MDRAID_ALIGN_BYTES	8*512	//(should be divisible by 8 sectors to keep 4kB alignment)
+
+static time_t mdraid_time = 0; //Array creation timestamp has to be identical across all the raid members, so we share it between invocations
```
This is done on purpose to ensure all arrays (and disks) created in single genimage run will have same creation time, there is nothing wrong with multiple arrays sharing timestamp (should use UUID to distinguish between them).

Also it is not required for member disks to have parent defined in configu, parent is mostly for convenience but you can still specify UUID manualy, in that case it's good to have times synced up unless time is manualy specified.

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+	//__u8  pad[256 - 136]; /* set to zero */
+
+	//Increase bitmap chunk size till we fit in sectors max
+	while(bsb->sectors_reserved > BITMAP_SECTORS_MAX) {
```
You seem to be very punctual about code formating. Have you considered providing .clang-format file in genimage, so we don't have to go through this again and again?

michaelolbrich replied:
Good point. I've started some time ago and forgot it again. I'll clean it up and create a PR.

michaelolbrich replied:
I won't merge that for now, because it conflicts with your changes. So please just copy the .clang-format from there for now.

Harvie replied:
Ok thanks, i'll mark all formating related stuff as resolved.

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+
+		/* constant array information - 128 bytes */
+		sb->magic = MD_SB_MAGIC;	/* MD_SB_MAGIC: 0xa92b4efc - little endian. */
+		sb->major_version = 1;	/* 1 */
```
Actualy most of the comments are copypasted from struct definition in kernel headers... I've just added couple informations that seemed important while i was writing the code.

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+		/* constant array information - 128 bytes */
+		sb->magic = MD_SB_MAGIC;	/* MD_SB_MAGIC: 0xa92b4efc - little endian. */
+		sb->major_version = 1;	/* 1 */
+		sb->feature_map = MD_FEATURE_BITMAP_OFFSET;	/* bit 0 set if 'bitmap_offset' is meaningful */
```
(again just copypasted from kernel as is, i honestly consider code more readable this way, but if you insist, why not take it one step further and go for doxygen compatible syntax? which btw allows comments at the end of line...)

michaelolbrich replied:
It's mostly about the line length. While I don't enforce a strict limit, I prefer to keep it below 100 chars when reasonably possible.

Also, while the comments in the header make sense like that, because they are aligned, here they are a big mess because with the added assignment, the alignment is gone.

Harvie replied:
Aligned

Harvie commented:
> And please squash all the mdraid commits.

Is it OK if you just click squash and merge instead of merge?

![image](https://github.com/user-attachments/assets/75c61974-22f3-42a8-9e60-70a2fc19aa00)

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+	 * have a meaningful role.
+	 */
+	__le16	*dev_roles = (__le16 *) ((char *) sb + sizeof(struct mdp_superblock_1)); /* role in array, or 0xffff for a spare, or 0xfffe for faulty */
+	//memset(dev_roles, 0xFF, max_devices*2); //All devices in array are set as inactive initialy
```
What do you mean by "zero of 0xFF" ?

Harvie replied:
I've commented that out, because i no longer build the array in a way where it would contain inactive members. Left it out in case i wanna support adding inactive members in future, but it's not needed now, because now i assign roles to all members.

Harvie replied:
This is also one of the reasons i like comments being at the end of the line, because it's immediately obvious that the comment should be ignored, since whole line is commented out.

Harvie commented on image-mdraid.c:
```diff
@@ -0,0 +1,414 @@
+		return 5;
+	}
+
+	//MD Superblock and Bitmap Superblock
```
Can do that. Why do people prefer those tho? Never understood that. If i write block with bunch of /* */ i cannot later comment out whole block using /* */ so it makes temporary disabling of whole code blocks much harder to do (unless you use fancy IDE which can figure that out for you - but then you can only uncomment that block using the same IDE).

Original idea was that code function should not realy too much on newlines. But is that real usecase? It's not like i am gonna remove all newlines in preprocessor before compiling? Are you really that concerned about being able to compile genimage in compiler from 1980s which does not support single line comments? Because in that case i would be much more concerned about all the other newfangled C99 stuff that is used in genimage... ;-)

Harvie replied:
OK, unified and aligned

michaelolbrich commented:
> > And please squash all the mdraid commits.
> 
> Is it OK if you just click squash and merge instead of merge?

Hmm, I thought the F2FS stuff was already a separate commit but it's not. That should really be separate. If you create a separate PR for that and remove it here, then I can do that (hopefully I won't forget that... :-))

Harvie commented:
> Hmm, I thought the F2FS stuff was already a separate commit but it's not. That should really be separate.

OK, squashed this PR into two commits. MDRAID and F2FS.

michaelolbrich commented:
At the end of the day, I'm the one that needs to maintain this code. And while I understand, that you prefer a different coding style, than we have in genimage so far, consistency is important to me.

But I really don't want to argue another 10 rounds about style issues, so I'll merge this and clean it up myself.

Harvie commented:
> At the end of the day, I'm the one that needs to maintain this code.
> But I really don't want to argue another 10 rounds about style issues, so I'll merge this and clean it up myself.

Thanks. I just hope there are no hard feelings :innocent: i'd still like to continue the development :smile: 
I really hope in the future we'll be able to find some reasonable .clang-format to cover all the cases.
I've still some issues with all configurations i've tried. I works perfectly for like 95% of the code, but then the last 5% is like totaly unacceptable for me in terms of readability.

Maybe i'm gonna check what they do in Linux:
https://github.com/torvalds/linux/blob/master/.clang-format
Haven't tried it yet on any project outside of kernel...
